libtesting-ui/src/main/java/com/mapbox/navigation/testing/ui/http/MockWebServerRule.kt

```diff
@@ -0,0 +1,68 @@
+package com.mapbox.navigation.testing.ui.http
+
+import okhttp3.mockwebserver.Dispatcher
+import okhttp3.mockwebserver.MockResponse
+import okhttp3.mockwebserver.MockWebServer
+import okhttp3.mockwebserver.RecordedRequest
+import org.junit.rules.TestWatcher
+import org.junit.runner.Description
+
+/**
+ * Creates and initializes a [MockWebServer] for each test.
+ *
+ * All request that should be mocked need to have a base url equal to [baseUrl]
+ * and have a [MockRequestHandler] that can serve the request added to the [requestHandlers] list.
+ */
+class MockWebServerRule : TestWatcher() {
+
+    val webServer = MockWebServer()
+
+    /**
+     * @see [MockWebServer.url]
+     */
+    val baseUrl = webServer.url("").toString()
+
+    /**
+     * Add [MockRequestHandler]s to this list for each request that should be handled.
+     *
+     * All mocked request should target [baseUrl].
+     *
+     * Cleared after each test.
+     */
+    val requestHandlers = mutableListOf<MockRequestHandler>()
+
+    init {
+        webServer.setDispatcher(object : Dispatcher() {
+            override fun dispatch(request: RecordedRequest): MockResponse {
+                requestHandlers.forEach {
+                    it.handle(request)?.run { return this }
+                }
+                throw Error(
+                    """request url:
+                      |${request.path}
+                      |is not handled
+                    """.trimMargin()
+                )
+            }
+        })
+    }
+
+    override fun starting(description: Description?) {
+        // no need to start the webServer, already started implicitly when fetching baseUrl
+        // webServer.start()
+    }
+
+    override fun finished(description: Description?) {
+        requestHandlers.clear()
+        webServer.shutdown()
+    }
+}
+
+/*val handshaCertificate = HandshakeCertificates.Builder()
+            .addPlatformTrustedCertificates()
+            .addTrustedCertificate()
+        val localhostCertificate = HeldCertificate.Builder()
+            .addSubjectAlternativeName(InetAddress.getByName("localhost").canonicalHostName)
+            .build()*/
+// webServer.useHttps(SSLContext.getDefault().socketFactory, false)
+// webServer.useHttps(TlsUtil.localhost().sslSocketFactory(), false)
```

Comment:
NIT Remove?

Reply:
I left it to reimplement when we switch to OkHttp 4. But you might be right, we can remove and rediscover when we need to.